File: wallpaper.js
// wallpaper.js
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');
const axios = require('axios');
const wallpaper = require('wallpaper');

const os = require('os');

class WallpaperChanger {
    async setWallpaper(filePath) {
        try {
            const absPath = path.resolve(filePath);
            await wallpaper.set(absPath);
            return true;
        } catch (err) {
            console.error('Failed to set wallpaper:', err.message);
            return false;
        }
    }

    async getRandomImage(folder) {
        const exts = ['.jpg', '.jpeg', '.png', '.bmp', '.gif', '.webp'];
        const files = fs.readdirSync(folder)
            .filter(f => exts.includes(path.extname(f).toLowerCase()))
            .map(f => path.join(folder, f));
        if (files.length === 0) return null;
        return files[Math.floor(Math.random() * files.length)];
    }

    async downloadUnsplash() {
        try {
            const response = await axios({
                method: 'get',
                url: 'https://source.unsplash.com/random/1920x1080',
                responseType: 'stream'
            });
            const tmpFile = path.join(os.tmpdir(), `wallpaper_${Date.now()}.jpg`);
            const writer = fs.createWriteStream(tmpFile);
            response.data.pipe(writer);
            return new Promise((resolve, reject) => {
                writer.on('finish', () => resolve(tmpFile));
                writer.on('error', reject);
            });
        } catch (err) {
            console.error('Download failed:', err.message);
            return null;
        }
    }

    async runSlideshow(folder, interval) {
        console.log(`Starting slideshow from ${folder}, interval ${interval}s`);
        setInterval(async () => {
            const img = await this.getRandomImage(folder);
            if (img) {
                const ok = await this.setWallpaper(img);
                if (ok) console.log(`Set: ${img}`);
            }
        }, interval * 1000);
    }
}

async function main() {
    const args = process.argv.slice(2);
    let file = null, folder = null, interval = 0, unsplash = false;
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--file' || args[i] === '-f') { file = args[++i]; }
        else if (args[i] === '--folder' || args[i] === '-d') { folder = args[++i]; }
        else if (args[i] === '--interval' || args[i] === '-i') { interval = parseInt(args[++i], 10); }
        else if (args[i] === '--unsplash' || args[i] === '-u') { unsplash = true; }
        else if (args[i] === '--help' || args[i] === '-h') {
            console.log('Usage: node wallpaper.js --file <file> | --folder <dir> [--interval <sec>] | --unsplash');
            process.exit(0);
        }
    }

    const changer = new WallpaperChanger();

    if (file) {
        await changer.setWallpaper(file);
    } else if (folder && interval > 0) {
        await changer.runSlideshow(folder, interval);
    } else if (folder) {
        const img = await changer.getRandomImage(folder);
        if (img) await changer.setWallpaper(img);
    } else if (unsplash) {
        const img = await changer.downloadUnsplash();
        if (img) {
            await changer.setWallpaper(img);
            fs.unlinkSync(img); // cleanup
        }
    } else {
        console.log('Usage: node wallpaper.js --file <file> | --folder <dir> [--interval <sec>] | --unsplash');
    }
}

main().catch(console.error);

File: wallpaper.swift
// wallpaper.swift
import Foundation
import AppKit

class WallpaperChanger {
    func setWallpaper(_ path: String) -> Bool {
        let fileURL = URL(fileURLWithPath: path)
        guard FileManager.default.fileExists(atPath: path) else {
            print("File not found.")
            return false
        }
        #if os(macOS)
        do {
            try NSWorkspace.shared.setDesktopImageURL(fileURL, forScreen: NSScreen.main!, options: [:])
            return true
        } catch {
            print("Failed to set wallpaper: \(error)")
            return false
        }
        #else
        // Linux or other: use system commands
        let abs = URL(fileURLWithPath: path).path
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        process.arguments = ["gsettings", "set", "org.gnome.desktop.background", "picture-uri", "file://"+abs]
        do {
            try process.run()
            process.waitUntilExit()
            return process.terminationStatus == 0
        } catch {
            // Try XFCE
            let process2 = Process()
            process2.executableURL = URL(fileURLWithPath: "/usr/bin/env")
            process2.arguments = ["xfconf-query", "-c", "xfce4-desktop", "-p", "/backdrop/screen0/monitor0/image-path", "-s", abs]
            do {
                try process2.run()
                process2.waitUntilExit()
                return process2.terminationStatus == 0
            } catch {
                return false
            }
        }
        #endif
    }

    func getRandomImage(from folder: String) -> String? {
        let exts = [".jpg", ".jpeg", ".png", ".bmp", ".gif", ".webp"]
        guard let enumerator = FileManager.default.enumerator(at: URL(fileURLWithPath: folder), includingPropertiesForKeys: nil) else {
            return nil
        }
        var files: [String] = []
        for case let fileURL as URL in enumerator {
            if fileURL.hasDirectoryPath { continue }
            let ext = fileURL.pathExtension.lowercased()
            if exts.contains("." + ext) {
                files.append(fileURL.path)
            }
        }
        return files.randomElement()
    }

    func downloadUnsplash() -> String? {
        guard let url = URL(string: "https://source.unsplash.com/random/1920x1080") else { return nil }
        let semaphore = DispatchSemaphore(value: 0)
        var result: String? = nil
        let task = URLSession.shared.dataTask(with: url) { data, response, error in
            if let data = data, let image = NSImage(data: data) {
                let tmp = NSTemporaryDirectory() + "wallpaper_\(Date().timeIntervalSince1970).jpg"
                if let tiff = image.tiffRepresentation,
                   let bitmap = NSBitmapImageRep(data: tiff),
                   let jpeg = bitmap.representation(using: .jpeg, properties: [:]) {
                    try? jpeg.write(to: URL(fileURLWithPath: tmp))
                    result = tmp
                }
            }
            semaphore.signal()
        }
        task.resume()
        semaphore.wait()
        return result
    }

    func slideshow(from folder: String, interval: Int) {
        print("Slideshow from \(folder), interval \(interval)s")
        while true {
            if let img = getRandomImage(from: folder) {
                if setWallpaper(img) {
                    print("Set: \(img)")
                }
            }
            Thread.sleep(forTimeInterval: TimeInterval(interval))
        }
    }
}

func main() {
    let args = CommandLine.arguments.dropFirst()
    var file: String? = nil
    var folder: String? = nil
    var interval: Int = 0
    var unsplash: Bool = false

    var i = args.startIndex
    while i < args.endIndex {
        let arg = args[i]
        switch arg {
        case "--file", "-f":
            if i+1 < args.endIndex { file = args[i+1]; i += 2 } else { i += 1 }
        case "--folder", "-d":
            if i+1 < args.endIndex { folder = args[i+1]; i += 2 } else { i += 1 }
        case "--interval", "-i":
            if i+1 < args.endIndex { interval = Int(args[i+1]) ?? 0; i += 2 } else { i += 1 }
        case "--unsplash", "-u":
            unsplash = true; i += 1
        case "--help", "-h":
            print("Usage: swift wallpaper.swift --file <file> | --folder <dir> [--interval <sec>] | --unsplash")
            return
        default:
            i += 1
        }
    }

    let changer = WallpaperChanger()

    if let f = file {
        _ = changer.setWallpaper(f)
    } else if let f = folder, interval > 0 {
        changer.slideshow(from: f, interval: interval)
    } else if let f = folder {
        if let img = changer.getRandomImage(from: f) {
            _ = changer.setWallpaper(img)
        }
    } else if unsplash {
        if let img = changer.downloadUnsplash() {
            _ = changer.setWallpaper(img)
            try? FileManager.default.removeItem(atPath: img)
        }
    } else {
        print("Usage: swift wallpaper.swift --file <file> | --folder <dir> [--interval <sec>] | --unsplash")
    }
}

main()
